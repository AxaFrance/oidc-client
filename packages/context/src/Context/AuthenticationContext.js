import React from 'react';
import PropTypes from 'prop-types';
import { AuthenticationContext } from './AuthenticationContextCreator';
import { OidcRoutes } from '../Routes';

const propTypes = {
  notAuthentified: PropTypes.node,
  notAuthorized: PropTypes.node,
  isLoading: PropTypes.bool.isRequired,
  isEnabled: PropTypes.bool,
  // eslint-disable-next-line
  oidcUser: PropTypes.object, //TODO : rÃ©cuperer le proptypes depuis OIDC client ?
  error: PropTypes.string.isRequired,
  login: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
};

const defaultProps = {
  notAuthentified: null,
  notAuthorized: null,
  isEnabled: true,
};

const AuthenticationProviderComponent = ({
  isLoading,
  isEnabled,
  oidcUser,
  error,
  login,
  logout,
  notAuthentified,
  notAuthorized,
  children,
}) => (
  <AuthenticationContext.Provider
    value={{
      isLoading,
      oidcUser,
      error,
      login,
      logout,
      isEnabled,
    }}
  >
    <OidcRoutes notAuthentified={notAuthentified} notAuthorized={notAuthorized}>
      {children}
    </OidcRoutes>
  </AuthenticationContext.Provider>
);

AuthenticationProviderComponent.propTypes = propTypes;
AuthenticationProviderComponent.defaultProps = defaultProps;

export default AuthenticationProviderComponent;
