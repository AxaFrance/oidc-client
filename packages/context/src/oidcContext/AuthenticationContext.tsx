import React, { ComponentType } from 'react';
import { User, UserManagerEvents, UserManager } from 'oidc-client';

export type oidcContext = {
  oidcUser: User | null;
  isEnabled: boolean;
  login: Function;
  logout: Function;
  events: UserManagerEvents;
  authenticating: ComponentType;
  isLoading: boolean;
  userManager: UserManager;
  error: string;
};

export const AuthenticationContext = React.createContext<oidcContext>(null);

export const useReactOidc = () => {
  const { isEnabled, login, logout, oidcUser, events } = React.useContext(AuthenticationContext);
  return { isEnabled, login, logout, oidcUser, events };
};
