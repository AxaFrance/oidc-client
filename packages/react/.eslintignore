build
bin
node_modules
dist
**/*.spec.ts
**/*.spec.tsx
**/*.test.ts
**/*.test.tsx
**/*.stories.tsx
./src/setupTests.ts
setupTests.js
.eslintrc.js
config
jest.config.js
App.specold.tsx
