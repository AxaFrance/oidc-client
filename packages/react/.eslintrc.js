module.exports = {
    extends: ['./config/defaultEslintConfig'], 
    parserOptions: {
      project: './tsconfig.eslint.json',
      tsconfigRootDir: __dirname,
    },
    rules: {
      '@typescript-eslint/naming-convention': [
        'error',
        {
          'selector': 'variable',
          'types': ['boolean'],
          'format': ['PascalCase'],
          'prefix': ['is', 'with', 'should', 'has', 'can', 'did', 'will']
        }
      ]
    }
  }