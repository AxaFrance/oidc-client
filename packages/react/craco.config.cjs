module.exports = {
    webpack: {
      configure: (webpackConfig, { env, paths }) => {
        webpackConfig.resolve.extensionAlias = {
          ".js": [".ts", ".js", ".tsx", ".jsx"],
          ".mjs": [".mts", ".mjs"]
        }
        return webpackConfig;
      },
    },
  };