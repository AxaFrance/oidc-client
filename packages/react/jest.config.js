module.exports = {
    collectCoverage: true,
    collectCoverageFrom: ['src/**/*.{ts,tsx}'],
};