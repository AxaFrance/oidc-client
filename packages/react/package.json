{
  "name": "@axa-fr/react-oidc",
  "version": "5.7.3-alpha0",
  "private": false,
  "main": "dist/index.js",
  "jsnext:main": "dist/index.js",
  "description": "OpenID Connect & OAuth authentication using react",
  "repository": {
    "type": "git",
    "url": "https://github.com/AxaGuilDEv/react-oidc.git"
  },
  "files": [
    "dist",
    "src",
    "README.md",
    "tsconfig.json",
    "package.json",
    "package-lock.json"
  ],
  "keywords": [
    "oidc",
    "OpenID Connect",
    "openid",
    "oauth2",
    "oauth",
    "react"
  ],
  "scripts": {
    "copy": "copyfiles -f ./src/oidc/vanilla/OidcServiceWorker.js ./public && copyfiles -f -soft ./src/oidc/vanilla/OidcTrustedDomains.js ./public",
    "start": "npm run copy && cross-env PORT=4200 react-scripts start",
    "build": "npm i react react-dom && npm run copy && react-scripts build",
    "test": "react-scripts test --coverage",
    "eject": "react-scripts eject",
    "clean": "rimraf dist",
    "prepare": "npm run clean && tsc --build \"./tsconfig.json\" && copyfiles -f ./src/oidc/vanilla/OidcServiceWorker.js ./dist && copyfiles -f ./src/oidc/vanilla/OidcTrustedDomains.js ./dist"
  },
  "dependencies": {
    "@openid/appauth": "1.3.1"
  },
  "peerDependencies": {
    "react": "x",
    "react-dom": "x"
  },
  "devDependencies": {
    "react": "x",
    "react-dom": "x",
    "@testing-library/jest-dom": "5.16.4",
    "@testing-library/react": "13.1.1",
    "@testing-library/user-event": "14.1.1",
    "@types/react": "18.0.6",
    "bootstrap": "4.6.1",
    "copyfiles": "2.4.1",
    "cross-env": "^7.0.3",
    "cypress": "^9.5.0",
    "msw": "0.39.2",
    "react-router-dom": "5.3.0",
    "react-scripts": "5.0.1",
    "typescript": "4.6.3"
  },
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
