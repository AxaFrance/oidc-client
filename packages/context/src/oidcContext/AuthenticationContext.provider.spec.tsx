import * as React from 'react';

import { render, waitForDomChange, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { AuthenticationProviderInt, withComponentOverrideProps } from './AuthenticationContext.provider';
import { AuthenticationContext } from './AuthenticationContext';

const AppTest = () => {
  const { oidcUser, isLoading, error, isEnabled, login, logout } = React.useContext(AuthenticationContext);

  return (
    <>
      <span data-testid="oidcUser">{JSON.stringify(oidcUser)}</span>
      <span data-testid="isLoading">{JSON.stringify(isLoading)}</span>
      <span data-testid="error">{JSON.stringify(error)}</span>
      <span data-testid="isEnabled">{JSON.stringify(isEnabled)}</span>

      <button type="button" onClick={() => login()}>
        login
      </button>
      <button type="button" onClick={() => logout()}>
        logout
      </button>
    </>
  );
};

const userMock = {
  firstname: 'Jean',
  token: 'SQSDFsdqfsdf234AEFZF',
};

const userManagerMock = {
  getUser: jest.fn(() => Promise.resolve(userMock)),
  events: {
    addUserLoaded: jest.fn(),
    addSilentRenewError: jest.fn(),
    addUserUnloaded: jest.fn(),
    addUserSignedOut: jest.fn(),
    addAccessTokenExpired: jest.fn(),
    removeUserLoaded: jest.fn(),
    removeSilentRenewError: jest.fn(),
    removeUserUnloaded: jest.fn(),
    removeUserSignedOut: jest.fn(),
    removeAccessTokenExpired: jest.fn(),
  },
};

const propsMocks = {
  location: 'location',
  history: 'ReactOidcHistory',
  loggerLevel: 'loggerLevel',
  logger: 'loggerMock',
  notAuthenticated: () => {
    return <div>This is notAuthenticated</div>;
  },
  notAuthorized: () => {
    return <div>This is notAuthorized</div>;
  },
  authenticating: () => {
    return <div>This is authenticating</div>;
  },
  callbackComponentOverride: () => {
    return <div>This is callbackComponentOverride</div>;
  },
  sessionLostComponent: () => {
    return <div>This is sessionLostComponent</div>;
  },
  CallbackInt: () => {
    return <div>This is callback</div>;
  },
  UserStore: 'UserStore',
  isEnabled: true,
  configuration: 'configuration',
  authenticationServiceInt: () => userManagerMock,
  setLoggerInt: jest.fn(),
  OidcRoutesInt: jest.fn(({ children }) => {
    return <div>{children}</div>;
  }),
  oidcLogInt: {
    info: jest.fn(),
    debug: jest.fn(),
    error: jest.fn(),
  },
  authenticateUserInt: jest.fn(() => () => {}),
  logoutUserInt: jest.fn(),
};

const getWrapper = props => ({ children }) => <AuthenticationProviderInt {...props}>{children}</AuthenticationProviderInt>;

describe('AuthContext tests suite', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    propsMocks.logoutUserInt.mockImplementation(() => Promise.resolve());
  });

  it('should mount provider without error', async () => {
    const { asFragment } = render(<AppTest />, { wrapper: getWrapper(propsMocks) });
    // Should show loading in snap
    expect(asFragment()).toMatchSnapshot();
    await waitForDomChange();
    // Should show user and cancel loading in snap
    expect(asFragment()).toMatchSnapshot();
  });

  it('should mount provider without error without callbackComponentOverride', async () => {
    const { asFragment } = render(<AppTest />, {
      wrapper: getWrapper({ ...propsMocks, callbackComponentOverride: undefined }),
    });

    await waitForDomChange();
    // Should show user and cancel loading in snap
    expect(asFragment()).toMatchSnapshot();
  });

  it('should call corrects initializer on mount', async () => {
    render(<AppTest />, { wrapper: getWrapper(propsMocks) });
    await waitForDomChange();

    expect(propsMocks.setLoggerInt).toBeCalledWith('loggerLevel', 'loggerMock');
    expect(userManagerMock.events.addUserLoaded).toBeCalled();
    expect(userManagerMock.events.addSilentRenewError).toBeCalled();
    expect(userManagerMock.events.addUserUnloaded).toBeCalled();
    expect(userManagerMock.events.addUserSignedOut).toBeCalled();
    expect(userManagerMock.events.addAccessTokenExpired).toBeCalled();
    expect(userManagerMock.getUser).toBeCalled();
  });

  it('should not cll loead user if component unomount', async () => {
    const { getByTestId, unmount } = render(<AppTest />, { wrapper: getWrapper(propsMocks) });
    const userValue = getByTestId('oidcUser');
    unmount();
    expect(userValue).toHaveTextContent('null');
  });

  it('should remove events  on unmount', async () => {
    const { unmount } = render(<AppTest />, { wrapper: getWrapper(propsMocks) });
    await waitForDomChange();

    expect(userManagerMock.events.removeUserLoaded).not.toBeCalled();
    expect(userManagerMock.events.removeSilentRenewError).not.toBeCalled();
    expect(userManagerMock.events.removeUserUnloaded).not.toBeCalled();
    expect(userManagerMock.events.removeUserSignedOut).not.toBeCalled();
    expect(userManagerMock.events.removeAccessTokenExpired).not.toBeCalled();

    unmount();

    expect(userManagerMock.events.removeUserLoaded).toBeCalled();
    expect(userManagerMock.events.removeSilentRenewError).toBeCalled();
    expect(userManagerMock.events.removeUserUnloaded).toBeCalled();
    expect(userManagerMock.events.removeUserSignedOut).toBeCalled();
    expect(userManagerMock.events.removeAccessTokenExpired).toBeCalled();
  });

  it('should change state and call when click on login', async () => {
    const { getByTestId, getByText } = render(<AppTest />, { wrapper: getWrapper(propsMocks) });
    await waitForDomChange();
    const loginButton = getByText('login');
    const loading = getByTestId('isLoading');
    expect(loading).toHaveTextContent('false');

    fireEvent.click(loginButton);
    expect(propsMocks.authenticateUserInt).toBeCalledWith(userManagerMock, propsMocks.location, propsMocks.history);
    expect(loading).toHaveTextContent('true');
  });

  it('should change state and call when click on logout', async () => {
    const { getByText } = render(<AppTest />, { wrapper: getWrapper(propsMocks) });
    await waitForDomChange();
    const logoutButton = getByText('logout');

    fireEvent.click(logoutButton);
    expect(propsMocks.logoutUserInt).toBeCalledWith(userManagerMock);
  });

  it('should change state and call when click on logout', async () => {
    propsMocks.logoutUserInt.mockImplementation(() => Promise.reject(new Error('error occured 1233123')));
    const { getByTestId, getByText } = render(<AppTest />, { wrapper: getWrapper(propsMocks) });
    await waitForDomChange();
    const logoutButton = getByText('logout');
    const error = getByTestId('error');
    expect(error).toHaveTextContent('""');

    fireEvent.click(logoutButton);

    await waitForDomChange();

    expect(propsMocks.logoutUserInt).toBeCalledWith(userManagerMock);
    expect(error).toHaveTextContent('error occured 1233123');
  });

  it('should return correct component', () => {
    const ComponentCallback = (props: React.PropsWithChildren<any>) => {
      const { callbackComponentOverride: OverrideComponent, ...otherProps } = props;
      return (
        <div>
          <div>
            <OverrideComponent />
          </div>
          <span data-testid="oidcUser">{JSON.stringify(otherProps)}</span>
        </div>
      );
    };

    const propsHoc = {
      prop1: 'value1',
      prop2: 'value2',
    };

    const CustomCallback = (props: React.PropsWithChildren<any>) => <div>Custom callback</div>;

    const WithHoc = withComponentOverrideProps(ComponentCallback, CustomCallback);

    const { asFragment } = render(<WithHoc {...propsHoc} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
