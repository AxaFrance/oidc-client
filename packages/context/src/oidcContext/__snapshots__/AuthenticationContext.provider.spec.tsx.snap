// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthContext tests suite should mount provider without error 1`] = `
<DocumentFragment>
  <div>
    <span
      data-testid="oidcUser"
    >
      null
    </span>
    <span
      data-testid="isLoading"
    >
      true
    </span>
    <span
      data-testid="error"
    >
      ""
    </span>
    <span
      data-testid="isEnabled"
    >
      true
    </span>
    <button
      type="button"
    >
      login
    </button>
    <button
      type="button"
    >
      logout
    </button>
  </div>
</DocumentFragment>
`;

exports[`AuthContext tests suite should mount provider without error 2`] = `
<DocumentFragment>
  <div>
    <span
      data-testid="oidcUser"
    >
      {"firstname":"Jean","token":"SQSDFsdqfsdf234AEFZF"}
    </span>
    <span
      data-testid="isLoading"
    >
      false
    </span>
    <span
      data-testid="error"
    >
      ""
    </span>
    <span
      data-testid="isEnabled"
    >
      true
    </span>
    <button
      type="button"
    >
      login
    </button>
    <button
      type="button"
    >
      logout
    </button>
  </div>
</DocumentFragment>
`;

exports[`AuthContext tests suite should mount provider without error without callbackComponentOverride 1`] = `
<DocumentFragment>
  <div>
    <span
      data-testid="oidcUser"
    >
      {"firstname":"Jean","token":"SQSDFsdqfsdf234AEFZF"}
    </span>
    <span
      data-testid="isLoading"
    >
      false
    </span>
    <span
      data-testid="error"
    >
      ""
    </span>
    <span
      data-testid="isEnabled"
    >
      true
    </span>
    <button
      type="button"
    >
      login
    </button>
    <button
      type="button"
    >
      logout
    </button>
  </div>
</DocumentFragment>
`;

exports[`AuthContext tests suite should return correct component 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        Custom callback
      </div>
    </div>
    <span
      data-testid="oidcUser"
    >
      {"prop1":"value1","prop2":"value2"}
    </span>
  </div>
</DocumentFragment>
`;
